### PR TITLE
Extension point for package manager.

### PR DESCRIPTION
**Pull request by KevinRansom**
Is the package manager extension point from: https://github.com/Microsoft/visualfsharp/pull/2483

1.  Does not contain the package manager.
2.  Builds on coreclr
3.  Requires package manager key to be at least 2 chars.  This allows references to be rooted by a drive letter.

### CODE REVIEW

File: tests/fsharpqa/Source/test.lst [hunk 1]
```diff
@@ -250,6 +250,7 @@ Misc01			EntryPoint
 Misc01			Globalization
 Misc01,NoMT		Import
 Misc01,NoMT		..\..\..\testsprivate\fsharpqa\Source\InteractiveSession\AssemblyLoading
+Misc01,NoMT		InteractiveSession\Paket
```
Comment (smoothdeveloper) [inline tests/fsharpqa/Source/test.lst]: this one should point to UnkonwDependencyManager right?

**Comment by forki** [comment]
Where would we need to put the packagemanager dll?

**Comment by dsyme** [comment]
Exciting!

File: src/fsharp/DependencyManager.Integration.fs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        | _ ->
+            let managers = RegisteredDependencyManagers()
+            match managers |> Seq.tryFind (fun kv -> path.StartsWith(kv.Value.Key + ":" )) with
+            | None ->
+                errorR(createPackageManagerUnknownError (path.Split(':').[0]) m)
+                ReferenceType.UnknownType
+            | Some kv -> ReferenceType.RegisteredDependencyManager kv.Value
+    with
+    | e ->
+        errorR(Error(FSComp.SR.packageManagerError(e.Message),m))
```
Comment (smoothdeveloper) [inline src/fsharp/DependencyManager.Integration.fs]: Maybe it wouldn't hurt to have the full stack trace?

File: src/fsharp/DependencyManager.Integration.fs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    Error(FSComp.SR.packageManagerUnknown(packageManagerKey, String.Join(", ", searchPaths), registeredKeys),m)
+
+let tryFindDependencyManagerInPath m (path:string) : ReferenceType =
+    try
+        match path.IndexOf(":") with
+        | -1 | 1 ->
+            ReferenceType.Library path
+        | _ ->
+            let managers = RegisteredDependencyManagers()
+            match managers |> Seq.tryFind (fun kv -> path.StartsWith(kv.Value.Key + ":" )) with
```
Comment (smoothdeveloper) [inline src/fsharp/DependencyManager.Integration.fs]: We need to fix the lookup, didn't had chance to fix that in original PR but we are using a map, so we should lookup on the exact key rather than iteration.

**Comment by KevinRansom** [comment]
@forki, currently, the directory containing the executing FSharp.Compiler.dll, or the BaseDirectory of the AppDomain or LoadContext, I.e the directory containing fsi.exe.

**Comment by forki** [comment]
I think that would not work for Fsi in VS. We can't guess that path from
paket install. We should decide to scan in some specific appdata path.

Am 21.03.2017 16:21 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by smoothdeveloper** [comment]
@forki / @KevinRansom there are few notes on that topic on the RFC: https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1027-fsi-references.md look for "Handler resolution" on it.

**Comment by forki** [comment]
How can we move forward here?

**Comment by smoothdeveloper** [comment]
@KevinRansom, I think we should keep some more iteration on @forki's branch, because having a paket-less version of this branch makes it harder to do real things with it, and we are losing important tests that found real issues along the way.

I might resume the work some time next week to sketch an implementation for the "handler resolution", and I think working on the other branch makes most sense, what about you submitting patches to the other branch instead?

Also, I think it would make sense for FSI to have the paket manager assembly since there is no hard dependency, just relying on paket.exe. The handler resolution also allow to override handler that we would be shipping out of the box.

Any input?

**Comment by KevinRansom** [comment]
By all means, forki’s branch should be where you guys drive this feature.  I need this so that I can start work on what it means for us.  I would caution you to resolve the handler in as simple a manner as possible.  We can add complexity, once we have some solid experience of the limitations of the most simplest scheme on multiple platforms in various scenarios.

**Comment by smoothdeveloper** [comment]
@KevinRansom can you review the https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1027-fsi-references.md "Handler resolution" section to tell if you think it is too complicated and what you propose instead?

**Comment by KevinRansom** [comment]
I'm not sure I understand it, I got a bit confused between tool resolution and assembly resolution I thought I did then I lost my place, and sort of got confused.

I think that the rules for #r assembly resolution can be quite simple though.

__CAVEAT:  The current Windows FSI #r behavior needs to be unchanged on Windows, which makes me wonder about what follows.  Perhaps a #newloader pragma to turn on this behavior__

1.  for an unqualified assembly name or path without a directory specification 
     E.g.
        "System.Runtime.Collections" or "System.Runtime.Collections.dll" 

    Load it using the standard Loader semantic.  I.e relative to the BaseDirectory of the FSI loadContext.

2.    For an absolute or relative path to the assembly Load it using the standard loader semantic relative to the script directory of the currently executing FSX.  If there is no executing script then relative to the fsi.exe location.  (Send to FSI needs to specify the executing script location, or project location).

3.  Resolutions using paket or nuget specify the full location of the file on disk.

4.  There is no longer any need to distinguish between ref and impl assemblies.  paket or nuget will need to ensure that they add the impl assemblies in the script before the ref assemblies.

5.  For resolving the tool, I think loadfrom files in an environment variable perhaps FSHARP_DM that has a semicolon delimited list of the full path to the location perhaps containing assemblies with the name pattern "*DependencyManager*.dll", if it's not found there then load from base directory.   For sure we will ship paket and nuget integration points in the product so they will be there.  And I suppose if someone builds one an NPM integration point, so the environment variable will likely almost never be used

Kevin

**Comment by KevinRansom** [comment]
@smoothdeveloper BTW, the reason I hadn't pulled this was I was hoping to build a test that demonstrated that it worked, but got sidetracked by PDB work..

File: src/fsharp/DependencyManager.Integration.fs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+let tryFindDependencyManagerByKey m (key:string) : IDependencyManagerProvider option =
+    try
+        RegisteredDependencyManagers() |> Map.tryFind key
+    with 
+    | e -> 
+        errorR(Error(FSComp.SR.packageManagerError(e.Message),m))
+        None
+
+let resolve (packageManager:IDependencyManagerProvider) implicitIncludeDir fileName m packageManagerTextLines =
```
Comment (dsyme) [inline src/fsharp/DependencyManager.Integration.fs]: Could we have /// comments on all new top-level ``let`` , ``type`` and ``module`` please?  thanks

**Comment by dsyme** [comment]
I glanced over this, it's looking very good.  What's the plan for adding more testing here?  There's only one small negative test so far, and the feature PR should really be self-contained w.r.t. testing?

Also is there any testing for the ref/impl split?

File: src/fsharp/CompileOps.fs [hunk 5]
```diff
@@ -4910,17 +4975,21 @@ module private ScriptPreprocessClosure =
...
                             let tcConfigResult, noWarns = ApplyMetaCommandsFromInputToTcConfigAndGatherNoWarn (!tcConfig, parsedScriptAst, pathOfMetaCommandSource)
                             tcConfig := tcConfigResult // We accumulate the tcConfig in order to collect assembly references
-                        
+                                                        
+                            yield! resolveDependencyManagerSources filename
+
                             let postSources = (!tcConfig).GetAvailableLoadedSources()
                             let sources = if preSources.Length < postSources.Length then postSources.[preSources.Length..] else []
 
+                            yield! resolveDependencyManagerSources filename
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: Why is ``resolveDependencyManagerSources`` called twice?

File: src/fsharp/DependencyManager.Integration.fs [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+#if FX_RESHAPED_REFLECTION
+open Microsoft.FSharp.Core.ReflectionAdapters
+#endif
+
+// NOTE: this contains mostly members whose intents are :
+// * to keep ReferenceLoading.PaketHandler usable outside of F# (so it can be used in scriptcs & others)
+// * to minimize footprint of integration in fsi/CompileOps
+
+/// hardcoded to net461 as we don't have fsi on netcore
+let targetFramework = "net461"
```
Comment (dsyme) [inline src/fsharp/DependencyManager.Integration.fs]: What's this, and does it need to be hard-coded?

File: src/fsharp/DependencyManager.Integration.fs (hunk 2 above)
Comment (dsyme) [inline src/fsharp/DependencyManager.Integration.fs]: Yeah adding some stack trace would be helpful

**Comment by dsyme** [comment]
@forki Can you update this when you get the chance?  Thanks!

**Comment by forki** [comment]
? I had my own pull request. Not sure what Kevin did here

**Comment by dsyme** [comment]
Closing in favour of https://github.com/Microsoft/visualfsharp/pull/4042

**Comment by KevinRansom** [comment]
I would like to keep this open for reference
